tony-core/src/main/java/com/linkedin/tony/horovod/HorovodDriver.java

```diff
@@ -0,0 +1,276 @@
+
+    /**
+     * @return Pair, left is Rendezvous server port, right is SlotInfo.
+     * @throws IOException
+     * @param taskProcess
+     */
+    private static Pair<Integer, List<SlotInfo>> waitTillServerStarted(final Process taskProcess) throws Exception {
+        assert DRIVER_SCRIPT_PATH != null;
+        Path parentPath = DRIVER_SCRIPT_PATH.getParent();
+        assert parentPath != null;
+
+        int checkCount = 0;
+        int maxCheckCount = 5;
+        Duration checkInterval = Duration.ofSeconds(2);
+
+        while (!existPortFile(parentPath) && (checkCount++) < maxCheckCount) {
+            if (taskProcess != null && !taskProcess.isAlive()) {
+                throw new Exception("Horovod Driver python process has finished, exit code: " + taskProcess.exitValue());
+            }
+
+            try {
+                LOG.info("Rendezvous server don't start, sleep for " + checkInterval.getSeconds() + " secs");
+                Thread.sleep(checkInterval.toMillis());
+            } catch (Exception e) {
+                LOG.warn(e);
+            }
+        }
+
+        if (checkCount > maxCheckCount) {
+            LOG.error("Timeout of starting horovod driver.");
+            throw new Exception("Errors on starting horovod driver within the fixed time.");
+        }
+
+        if (taskProcess != null && !taskProcess.isAlive()) {
+            String msg = "Driver python process has ended abnormally, exit code: " + taskProcess.exitValue();
+            LOG.error(msg);
+            throw new Exception(msg);
+        }
+        return getServerInfo(parentPath);
+    }
+
+    @VisibleForTesting
+    protected static Pair<Integer, List<SlotInfo>> getServerInfo(Path parentPath) throws IOException {
+        int port = -1;
+        File parentFile = parentPath.toFile();
+        requireNonNull(parentFile);
+        File[] files = parentFile.listFiles();
+        if (files == null) {
+            return Pair.of(port, null);
+        }
+
+        for (File file : files) {
+            String fileName = file.getName();
+            if (fileName.endsWith(PORT_FILE_NAME_SUFFIX)) {
+                int tempIndex = fileName.indexOf(PORT_FILE_NAME_SUFFIX);
+                port = Integer.parseInt(fileName.substring(0, tempIndex));
+                String fileContent = FileUtils.readFileToString(file);
+                // TODO: 4/10/21 fast fail when file content is empty.
+                LOG.info("Horovod rendezvous server slot info: \n" + fileContent);
+                List<SlotInfo> slotInfoList = new Gson().fromJson(fileContent,
+                        new TypeToken<List<SlotInfo>>() { }.getType());
+                return Pair.of(port, slotInfoList);
+            }
+        }
+        LOG.info("Still no starting horovod rendezvous server.");
+        return Pair.of(port, null);
+    }
+
+    private static boolean existPortFile(Path parentPath) throws IOException {
+        return getServerInfo(parentPath).getLeft() != -1 ? true : false;
```

Comment:
will there be a race condition on port allocation if there are a horovod job and a non-horovod job run in parallel?

Reply:
This won't happen. The port files will be created in independent application folder.

Reply:
Not file conflict. I meant port conflict. We had this issue before that if you launch multiple TF jobs on the same node, there might be a race condition. The problem was, we reserve a port in Java for the job, and later release it before the job starts. However, the job might not start right away and would take a few minutes to set up. During that few minutes, another TF job could take up that port and causes port conflict.

Reply:
It won't happen. In horovod driver, It is Python process(built-in horovod_driver.py) that selects an available port, which is not TonY reserved. Therefore, port race condition will not occur.

Reply:
Got it, makes sense